rattan-core/src/cells/token_bucket.rs

```diff
@@ -0,0 +1,1160 @@
+                                            self.packet_queue.enqueue(new_packet);
+                                            break;
+                                        } else {
+                                            self.packet_queue.enqueue(new_packet);
+                                        }
+
+                                    }
+                                }
+                            }
+                        }
+                        None => {
+                            // channel closed
+                            return None;
+                        }
+                    }
+                }
+                _ = self.timer.sleep(self.next_available - Instant::now()) => {
+                    break;
+                }
+            }
+        }
+
+        // There will be at least one packet in the queue
+        // get number of tokens added from last_check
+        let now = Instant::now();
+        debug!(
+            "Previous next_available distance: {:?}",
+            self.next_available - now
+        );
+        let burst_duration =
+            length_to_time(self.token_bucket.burst.as_u64(), self.token_bucket.rate);
+        let minburst_duration = length_to_time(
+            self.peak_token_bucket.burst.as_u64(),
+            self.peak_token_bucket.rate,
+        );
+        self.token_bucket.tokens = min(
+            burst_duration,
+            now - self.last_check + self.token_bucket.tokens,
+        );
+        self.peak_token_bucket.tokens = min(
+            minburst_duration,
+            now - self.last_check + self.peak_token_bucket.tokens,
+        );
+
+        let size = self.packet_queue.get_front_size().unwrap() as u64;
+        debug!("size: {}", size);
+        debug!("max_size: {}", self.max_size);
+        let consumed_tokens = length_to_time(size, self.token_bucket.rate);
+        let consumed_ptokens = length_to_time(size, self.peak_token_bucket.rate);
+
+        self.last_check = now;
+
+        // if tokens and ptokens are all positive, send the packet, and update number of tokens
+        if self.token_bucket.tokens >= consumed_tokens
+            && self.peak_token_bucket.tokens >= consumed_ptokens
+        {
+            let packet_to_send = self.packet_queue.dequeue().unwrap();
+            self.token_bucket.tokens -= consumed_tokens;
+            self.peak_token_bucket.tokens -= consumed_ptokens;
+            debug!(
+                "in dequeue, toks: {}, ptoks: {}",
+                self.token_bucket.tokens.as_nanos(),
+                self.peak_token_bucket.tokens.as_nanos()
+            );
+            // set next_available
+            if self.packet_queue.is_empty() {
+                self.next_available = now + LARGE_DURATION;
+            } else {
+                let next_size = self.packet_queue.get_front_size().unwrap() as u64;
+                self.next_available = self.calculate_next_available(now, ByteSize::b(next_size));
+            }
```

Comment:
Can be changed to:

```rust
if let Some(next_size) = self.packet_queue.get_fron_size {
    self.next_available = self.calculate_next_available(now, ByteSize::b(next_size as u64));
} else {
    self.next_available = now + LARGE_DURATION;
}
```

PS: Always remember to avoid unwrap. Even if your sanity check can ensure we have packets in the queue (the unwrap will be safe), we should reduce possible error spaces for future maintenance.